go/stats/counters.go

```diff
@@ -19,36 +19,95 @@ package stats
 import (
 	"bytes"
 	"fmt"
+	"strconv"
 	"strings"
 	"sync"
 	"sync/atomic"
+
+	log "github.com/golang/glog"
+
+	"vitess.io/vitess/go/sync2"
 )
 
+// Counter is expvar.Int+Get+hook
+type Counter struct {
+	i    sync2.AtomicInt64
+	help string
+}
+
+// NewCounter returns a new Counter
+func NewCounter(name string, help string) *Counter {
+	v := &Counter{help: help}
+	if name != "" {
+		publish(name, v)
+	}
+	return v
+}
+
+// Add adds the provided value to the Counter
+func (v *Counter) Add(delta int64) {
+	if delta < 0 {
+		log.Warningf("Adding a negative value to a counter, %v should be a gauge instead", v)
```

Comment:
How likely is it that this will get called?

If it's not likely at all because you already checked all callers in open-source, then I'm okay with it as is.

Otherwise, it may be a good idea to use a throttled logger here instead. This way we would avoid that bad code spams the log, makes the disk full and potentially slows down the process.

Reply:
Oh, that's a good idea. I've checked all the easy sites (`Add(-1)`) but there's a lot more that wind their way through the code that I don't have confidence on, hence logging for now instead of banning decrementing all together. I switched to using a throttled logger here. I also switched to using a throttled logger for the unsupported export type logger in `prometheusbackend.go`, thanks for the suggestion!